be/src/storage/tablet_manager.cpp

```diff
@@ -1182,21 +1168,42 @@ Status TabletManager::_drop_tablet_directly_unlocked(TTabletId tablet_id, bool k
     TabletSharedPtr dropped_tablet = it->second;
     tablet_map.erase(it);
     _remove_tablet_from_partition(*dropped_tablet);
-    if (!keep_state) {
-        LOG(INFO) << "Shutting down tablet " << tablet_id;
-        // drop tablet will update tablet meta, should lock
-        std::unique_lock wrlock(dropped_tablet->get_header_lock());
-        // NOTE: has to update tablet here, but must not update tablet meta directly.
-        // because other thread may hold the tablet object, they may save meta too.
-        // If update meta directly here, other thread may override the meta
-        // and the tablet will be loaded at restart time.
-        // To avoid this exception, we first set the state of the tablet to `SHUTDOWN`.
-        dropped_tablet->set_tablet_state(TABLET_SHUTDOWN);
-        dropped_tablet->save_meta();
+
+    DroppedTabletInfo drop_info{.tablet = dropped_tablet, .flag = flag};
+
+    if (flag == kDeleteFiles) {
         {
-            std::unique_lock wlock(_shutdown_tablets_lock);
-            _shutdown_tablets.emplace(dropped_tablet->tablet_id(), dropped_tablet);
+            // NOTE: Other threads may save the tablet meta back to storage again after we
+            // have deleted it here, and the tablet will reappear after restarted.
+            // To prevent this, set the tablet state to `SHUTDOWN` first before removing tablet
+            // meta from storage, and assuming that no thread will change the tablet state back
+            // to 'RUNNING' from 'SHUTDOWN'.
+            std::unique_lock l(dropped_tablet->get_header_lock());
+            dropped_tablet->set_tablet_state(TABLET_SHUTDOWN);
         }
+
+        // Remove tablet meta from storage, crash the program if failed.
+        if (auto st = _remove_tablet_meta(dropped_tablet); !st.ok()) {
+            LOG(FATAL) << "Fail to remove tablet meta: " << st;
```

Comment:
LOG(FATAL) will cause process crash, IMO returning error is a better option.

Reply:
> LOG(FATAL) will cause process crash, IMO returning error is a better option.

It's designed to be compatible with the original implementation.

Reply:
> > LOG(FATAL) will cause process crash, IMO returning error is a better option.
> 
> It's designed to be compatible with the original implementation: `TabletMeta::save_meta` will crash the process on error.

I can submit another PR to improve this.